### ISSUE TITLE
add citekey variable

### ISSUE THREAD

**Issue by denismaier** [citation-style-language/schema#243]
**Is your feature request related to a problem? Please describe.**
Switching from word-processors to plain text based systems (pandoc, latex) is currently problematic since CSL currently doesn't know a user-definable citekey.

Additional Details: https://discourse.citationstyles.org/t/citekey-variable/1651/32

**Describe the solution you'd like**
Suggestion per @bwiernik: 
An option could be to add a citekey variable to the CSL-JSON spec, with the indication that it could be used to in the same manner as a BibTeX citekey with id as a fallback if citekey is unavailable.

That would involve primarily minor changes with two players:

- pandoc adding citekey as primary source for the citation key. This would be easy enough as existing documents relying on id would still work.
- citation.js switching citation-label to citekey, which they should anyway.

**Comment by bdarcus** [comment]
@jgm - any feedback on this?

**Comment by jgm** [comment]
I only had time to scan the discussion briefly.  I'm still a bit unclear as to why people would want an id that was distinct from the citekey, but if people need this, it wouldn't be too complicated to change pandoc-citeproc to look up citations by citekey, falling back on id when citekey is not present.

**Comment by jgm** [comment]
Well, there would be some nomenclature issues, since pandoc's types have fields like `citationId`.  Changing these would be a breaking API change in pandoc-types, which I try to avoid as it requires changes throughout the ecosystem. Not changing them would lead to confusion.  So that's a bit of a drawback, but I wouldn't stand in the way if people need this.

**Comment by bwiernik** [comment]
I'm not sure what you mean about `citationId`. Where is that field used? In pandoc's internal data structure?

**Comment by denismaier** [comment]
`citationId` is used in citation objects in pandoc's ast.

**Comment by bdarcus** [comment]
I don't totally get it either @jgm, but I think the way to think about this is you keep citationId as is, but if you ingest csl-json that has a citekey variable, you use that instead of the id.

I think this idea comes from the world of GUI apps that work with Word and such, where id is often a machine-generated id.

**Comment by denismaier** [comment]
@adam3smith Perhaps you want to clarify? My own perspective on this is also more influenced by plain text systems...

**Comment by adam3smith** [comment]
Here are the two main points for distinguishing between id and citekey from the discourse thread:

1. there are already a large number of documents with existing, non-citekey ids in the embedded metadata (and while we _think_ they're not used for anything, we only _know_ this for Zotero and Mendeley)
2. some tools offer globally unique and stable IDs (such as Zotero URIs), whereas citekeys are only locally unique and ideally, but not necessarily, stable and it may be useful to keep a place for the former in the CSL data schema.

**Comment by bdarcus** [comment]
I'm going to move the discussion back here fron the PR, because I think we already have what we need; no need for new variables at all.

Please read [this thread from 2009](https://discourse.citationstyles.org/t/citation-label/761/6), which I think lays it out, with input from the citeproc-js and mendeley sides.

We need:

1. a place to put machine-readable and (maybe optional) human-readable ids on the input side; respectively, they are `id` and `citation-label`
2. a way to generate styling for the _output_ labels (aka citekeys); this is a separate, and solvable, problem, from the concern that motivated this issue

So say Zotero adds a new `citekey` field. If that's non-empty, then Zotero might output it like this (but in yaml):

``` yaml
id: 0fcce47a-b4ab-11ea-bebd-00d8616ef71c
citation-label: doe19
```
The CSL processor would use the label directly, or generate it from the metadata (as processors do now).

In short, we don't need a new json key or csl variable.

We do, perhaps, need some minor additions on the csl styling side, at least in the documentation. But that's a separate issue.

What am I missing?

**Comment by bwiernik** [comment]
@bdarcus No, `citekey` and `citation-label` are not the same thing at all.

Let me try to explain this again.
  - `citekey` is a human-readable, **input** code, used to identify the item in the input data. This is, for example, what a user of pandoc would type into their document. It is (at least locally-)unique and stable in the input data.
  - `citation-label` is a formatted **output** format, often automatically generated at rendering time. It is not necessarily stable (e.g., `Ferr1987` might become `Ferr1987a` if a second `Ferr1987` item is added to a document). These labels are often generated automatically following rules specified in a style.

They are completely different variables. `citation-label` needs to stay the same and not be changed.

What is at discussion is whether we need `citekey` in addition to `id`. There are two different ways of populating `id` currently in use. One uses `id` like the `citekey` in BibTeX (e.g., `@Ferr1987ShortTitle`)--a thing that users are expected to have to see and type in their document (this is how pandoc uses it). The other is as a machine-readable code (e.g., Zotero and Mendeley store the globally-unique item ID in `id` in their Word processor documents). There is potentially use for such a machine-readable `id` and (1) we don't know whether any programs rely on being able to use `id` in this way and (2) lots of existing data exists in the wild using `id` as a machine-readable identifier, rather than a human-usable citekey.

The proposed solution is a compromise that makes the least disruption possible.

**Comment by bdarcus** [comment]
> * `citekey` is a human-readable, **input** code, used to identify the item in the input data. This is, for example, what a user of pandoc would type into their document. It is (at least locally-)unique and stable in the input data.

Agreed.

> * `citation-label` is a formatted **output** format, often automatically generated at rendering time. It is not necessarily stable (e.g., `Ferr1987` might become `Ferr1987a` if a second `Ferr1987` item is added to a document). These labels are often generated automatically following rules specified in a style.

OK. 

But again, their only purpose in this context is to associate citation with reference in the context of a csl process and output. 

E.g they are, effectively, a human-readable id.

> They are completely different variables. citation-label needs to stay the same and not be changed.

This I don't understand. Why?

I also don't understand why the PR adds a new variable to `cs:style`.

**Comment by denismaier** [comment]
> I also don't understand why the PR adds a new variable to `cs:style`.

Where is that?

**Comment by bdarcus** [comment]
I use that as shorthand for the style schema; not the input json.

**Comment by denismaier** [comment]
I see.

**Comment by denismaier** [comment]
> I use that as shorthand for the style schema; not the input json.

So, see my anwer at the PR conversation.

**Comment by bwiernik** [comment]
> I also don't understand why the PR adds a new variable to cs:style.

As an option to use a CSL to effectively de-render a citation into its input format for a plain-text editor. That’s literally it. 

citation-label is an output format.  citekey is an input format. Making citekey available in output is _only_ for facilitating transfer between GUI and plaintext editing systems, which is otherwise hard to do currently.

**Comment by denismaier** [comment]
> As an option to use a CSL to effectively de-render a citation into its input format for a plain-text editor. That’s literally it.
> 
> citation-label is an output format. citekey is an input format. Making citekey available in output is _only_ for facilitating transfer between GUI and plaintext editing systems, which is otherwise hard to do currently.

So, if we had another solution for this transfer, we could do without `citekey` as a variable, right?

**Comment by bdarcus** [comment]
To be clear, I don't see a problem adding the key as a variable for output
styling. I was just asking.

>
